drivers/Serial.h

```diff
@@ -98,6 +98,8 @@ class Serial : public SerialBase, public Stream, private NonCopyable<Serial> {
     {
         return SerialBase::writeable();
     }
+    int getc();
```

Comment:
Why are we removing `virtual` for these methods? (Hint: use commit message to contain details for changes, in this case it would be a problem with having virtual getc/putc and how this change fixes the problem)

Reply:
the getc/putc in Serial class inherits from Stream class, and in Stream class, putc/getc are not virtual functions, so that's why I removed the virtual here. in Serial class.

Reply:
deriving an UnlockedSerial form Serial sounds just another version of switching to RawSerial and UARTSerial. @sg- Any suggestions?

Reply:
> deriving an UnlockedSerial form Serial sounds just another version of switching to RawSerial 

Well, yes. And actually it would be trickier than I was thinking, as it's not just the lock in the HAL class, like the other HAL classes, it's the fact that all the calls use the C library. Serial is designed as a C stream-based class.

`Serial::putc` is supposed to be "call fputc on my FileHandle", and as the C library is not callable from interrupt, `Serial::putc` cannot be used from interrupt. Similarly for `Serial::scanf`, etc.

If you want something that gives you simple getc and putc that are interrupt-safe, not using the C library, then that is `RawSerial`. As its header says:

    /** A serial port (UART) for communication with other serial devices
     * This is a variation of the Serial class that doesn't use streams,
     * thus making it safe to use in interrupt handlers with the RTOS.
     *

Your patch is an incomplete attempt to make `Serial` act like `RawSerial` if called from interrupt - "don't claim the mutex, and don't use the C library", but it only covers the API that `RawSerial` has. If that's all you're using, you may as well just use `RawSerial`, rather than attempt to bend `Serial`.

An `UnlockedSerial` would sort of work - it would let you access the `FileHandle` API, which would be useable from interrupt. You'd just have to avoid the C library-based API.

    class UnlockedSerial : public Serial {
          // override lock/unlock with empty functions
    }

    UnlockedSerial serial;
    // get past Stream's bizarre protection of its FileHandle base -
    // we can't call serial.read() directly
    FileHandle &fh =serial;
    n = fh.read(buffer, 1);  // only reading 1 is ever safe, if from IRQ - have to precheck that 1 with serial.readable().

In general `FileHandle`s are not expected to be useable from interrupt, but it happens that `Stream`'s implementation is, given an appropriate `lock`.

Not sure what that would really gain you though - it's just a very complicated way to keep using two horrible classes instead of the sensible `RawSerial` and `UARTSerial` though. Stream's broken FileHandle isn't very useful - eg the fact you have to read 1 above.